src/core/loss_detection.c

```diff
@@ -1496,6 +1498,69 @@ QuicLossDetectionProcessAckBlocks(
     }
 
     if (NewLargestAck) {
+        if (Connection->Settings.EcnEnabled && Path->EcnValidationState != ECN_VALIDATION_FAILED) {
+            //
+            // Per RFC 9000, we validate ECN counts from received ACK frames
+            // when the largest acked packet number increases.
+            //
+            BOOLEAN EcnValidated = TRUE;
+            int64_t EctCeDeltaSum = 0;
+            if (Ecn != NULL) {
+                EctCeDeltaSum +=
+                    Ecn->CE_Count - Connection->EcnCeCounters[EncryptLevel];
+                EctCeDeltaSum +=
+                    Ecn->ECT_0_Count - Connection->EcnEctCounters[EncryptLevel];
+                //
+                // Conditions where ECN validation fails:
+                // 1. Reneging ECN counts from the peer.
+                // 2. ECN counts do not match the marks that were applied to the packets sent.
+                //
+                if (EctCeDeltaSum < 0 ||
+                    EctCeDeltaSum < EcnEctCounter ||
+                    Ecn->ECT_1_Count != 0 ||
+                    Connection->NumPacketsSentWithEct < Ecn->ECT_0_Count) {
+                    EcnValidated = FALSE;
+                }
+
+                if (EcnValidated) {
+                    //
+                    // TODO: Notify CC of the ECN signal before we update the CE counts.
+                    //
+                    Connection->EcnCeCounters[EncryptLevel] = Ecn->CE_Count;
+                    Connection->EcnEctCounters[EncryptLevel] = Ecn->ECT_0_Count;
+                    if (Path->EcnValidationState <= ECN_VALIDATION_UNKNOWN) {
```

Comment:
Is this check necessary? It's just excluding FAILED, right? We're already in a conditional that says !FAILED. Just add a debug assert instead?

Reply:
This check is actually necessary. Otherwise, we would emit an ECN validation success log every time we receive a correctly marked packet.

Reply:
You resolve this, but didn't remove the check?